File: test/projectmanagement.js
const expect = require('chai').expect,
  conf = require('./conf'),
  WhmcsError = require('../lib/whmcserror');

describe('Module "Project Management"', function () {
  let demoProjectId;

  before(async function () {
    const opts = {
      title: 'demo project',
      adminid: 1
    };

    const res = await conf.whmcs.projectManagement.createProject(opts);
    demoProjectId = res.projectid;
  });

  it('should create a project', async function () {
    const opts = {
      title: 'untitled project',
      adminid: 1
    };
    const res = await conf.whmcs.projectManagement.createProject(opts);
    expect(res).to.have.a.property('projectid').to.not.be.null;
  });

  it('should get projects', async function () {
    const opts = {
      limitstart: 0,
      limitnum: 1
    };
    const res = await conf.whmcs.projectManagement.getProjects(opts);
    expect(res).to.have.a.property('numreturned').to.not.be.null;
    expect(res).to.have.a.property('projects').to.be.an.an('array').to.have.length.greaterThan(0);
  });

  it('should get project details by project id', async function () {
    const opts = {
      projectid: demoProjectId
    };
    const res = await conf.whmcs.projectManagement.getProject(opts);
    expect(res).to.have.a.property('projectinfo').to.be.an.an('object');
  });

  it('should update a project', async function () {
    const updateOpts = {
      projectid: demoProjectId,
      title: 'space oddity'
    };
    const updateRes = await conf.whmcs.projectManagement.updateProject(updateOpts);
    expect(updateRes).to.have.a.property('message').to.equal('Project Has Been Updated');

    const getOpts = {
      projectid: demoProjectId
    };
    const getRes = await conf.whmcs.projectManagement.getProject(getOpts);
    expect(getRes).to.have.a.property('projectinfo').to.be.an.an('object');
    expect(getRes.projectinfo).to.have.a.property('title').to.equal('space oddity');
  });

  it('should add a message to a project', async function () {
    const addOpts = {
      projectid: demoProjectId,
      message: 'can you hear me major tom?'
    };

    const addRes = await conf.whmcs.projectManagement.addProjectMessage(addOpts);
    expect(addRes).to.have.a.property('message').to.equal('Message has been added');

    const getOpts = {
      projectid: demoProjectId
    };
    const getRes = await conf.whmcs.projectManagement.getProject(getOpts);
    expect(getRes).to.have.a.property('messages').to.be.an.an('object');
    expect(getRes.messages).to.have.a.property('message').to.be.an('array').to.have.lengthOf(1);
    expect(getRes.messages.message[0]).to.have.a.property('message').to.equal('can you hear me major tom?');
  });

  it('should add a task to a project', async function () {
    const addOpts = {
      projectid: demoProjectId,
      duedate: '1969-07-11',
      task: 'leave the capsule'
    };

    const addRes = await conf.whmcs.projectManagement.addProjectTask(addOpts);
    expect(addRes).to.have.a.property('message').to.equal('Task has been added');


    const getOpts = {
      projectid: demoProjectId
    };
    const getRes = await conf.whmcs.projectManagement.getProject(getOpts);
    expect(getRes).to.have.a.property('tasks').to.be.an.an('object');
    expect(getRes.tasks).to.have.a.property('task').to.be.an('array').to.have.lengthOf(1);
    expect(getRes.tasks.task[0]).to.have.a.property('task').to.equal('leave the capsule');
    demoTaskId = getRes.tasks.task[0].id;
  });

  describe('Project task', function () {
    let demoTaskId;

    before(async function () {
      const addOpts = {
        projectid: demoProjectId,
        duedate: '1969-07-11',
        task: 'leave the capsule'
      };

      const addRes = await conf.whmcs.projectManagement.addProjectTask(addOpts);
      expect(addRes).to.have.a.property('message').to.equal('Task has been added');

      const getOpts = {
        projectid: demoProjectId
      };
      const getRes = await conf.whmcs.projectManagement.getProject(getOpts);
      expect(getRes).to.have.a.property('tasks').to.be.an.an('object');
      expect(getRes.tasks).to.have.a.property('task').to.be.an('array').to.have.length.greaterThan(0);
      expect(getRes.tasks.task[0]).to.have.a.property('task').to.equal('leave the capsule');
      demoTaskId = getRes.tasks.task[0].id;
    });

    it('should update a project task', async function () {
      const updateOpts = {
        taskid: demoTaskId,
        task: 'step through the door'
      };
      const updateRes = await conf.whmcs.projectManagement.updateProjectTask(updateOpts);
      expect(updateRes).to.have.a.property('message').to.equal('Task has been updated');

      const getOpts = {
        projectid: demoProjectId
      };
      const getRes = await conf.whmcs.projectManagement.getProject(getOpts);
      expect(getRes).to.have.a.property('tasks').to.be.an.an('object');
      expect(getRes.tasks).to.have.a.property('task').to.be.an('array').to.have.length.greaterThan(0);
      expect(getRes.tasks.task[0]).to.have.a.property('task').to.equal('step through the door');
    });

    it('should start and then end a project task timer', async function () {
      const timerOpts = {
        taskid: demoTaskId,
        projectid: demoProjectId
      };
      const timerRes = await conf.whmcs.projectManagement.startTaskTimer(timerOpts);
      expect(timerRes).to.have.a.property('message').to.equal('Start Timer Has Been Set');

      const getOpts = {
        projectid: demoProjectId
      };
      const getRes = await conf.whmcs.projectManagement.getProject(getOpts);
      expect(getRes).to.have.a.property('tasks').to.be.an.an('object');
      expect(getRes.tasks).to.have.a.property('task').to.be.an('array').to.have.length.greaterThan(0);
      expect(getRes.tasks.task[0]).to.have.a.property('timelogs').to.be.an.an('object');
      expect(getRes.tasks.task[0].timelogs).to.have.a.property('timelog').to.be.an.an('array').to.have.length.greaterThan(0);
      expect(getRes.tasks.task[0].timelogs.timelog[0]).to.have.a.property('id').to.not.be.null;
      demoTimerId = getRes.tasks.task[0].timelogs.timelog[0].id;

      const endOpts = {
        timerid: demoTimerId,
        projectid: demoProjectId
      };
      const endRes = await conf.whmcs.projectManagement.endTaskTimer(endOpts);
      expect(endRes).to.have.a.property('message').to.equal('Timer Has Ended');
    });

    it('should delete a project task', async function () {
      const deleteOpts = {
        projectid: demoProjectId,
        taskid: demoTaskId
      };
      const deleteRes = await conf.whmcs.projectManagement.deleteProjectTask(deleteOpts);
      expect(deleteRes).to.have.a.property('message').to.equal('Task has been deleted');
    });
  });
});

File: route/api/project_management.rb
namespace "/api/v1" do
  # GetProject
  get "/project" do
    response = Whmcs.call("GetProject", { :projectid => params[:projectid].to_i })
    logger.info '[#{response[:result].upcase}] - #{response[:message]}.'
    return json(response)
  end

  # GetProjects
  get "/projects" do
    response = Whmcs.call("GetProjects", { :limitstart => params[:limitstart].to_i || 0,
                                           :limitnum => params[:limitnum].to_i || 25,
                                           :userid => params[:userid].to_i,
                                           :title => params[:title].to_s,
                                           :ticketids => params[:ticketids].to_s,
                                           :invoiceids => params[:invoiceids].to_s,
                                           :notes => params[:notes].to_s,
                                           :adminid => params[:adminid].to_i,
                                           :status => params[:status].to_s,
                                           :createdfrom => params[:createdfrom].to_s,
                                           :duedate => params[:duedate].to_s,
                                           :completed => params[:completed],
                                           :lastmodified => params[:lastmodified].to_s })
    logger.info '[#{response[:result].upcase}] - #{response[:message]}.'
    return json(response)
  end

  # CreateProject
  post "/project/create" do
    # read request body
    payload = JSON.parse(request.body.read.to_s)

    logger.info "Create Project: #{payload}"
    response = Whmcs.call("CreateProject", { :title => payload[:title],
                                             :adminid => payload[:adminid].to_i,
                                             :userid => payload[:userid].to_i,
                                             :status => payload[:status],
                                             # The date of the project in Y-m-d format.
                                             :created => payload[:created].to_s,
                                             :duedate => payload[:duedate].to_s,
                                             :completed => payload[:completed],
                                             :ticketids => payload[:ticketids].to_s,
                                             :invoiceids => payload[:invoiceids].to_s })

    logger.info "Project Created:  #{response}"
    return json(response)
  end

  # UpdateProject
  post "/project/update" do
    payload = JSON.parse(request.body.read.to_s)

    logger.info "Update Project: #{payload}"
    response = Whmcs.call("UpdateProject", { :projectid => payload[:projectid].to_i,
                                             :adminid => payload[:adminid].to_i,
                                             :userid => payload[:userid].to_i,
                                             :status => payload[:status],
                                             # The date of the project in Y-m-d format.
                                             :created => payload[:created].to_s,
                                             :duedate => payload[:duedate].to_s,
                                             :completed => payload[:completed],
                                             :title => payload[:title],
                                             :ticketids => payload[:ticketids].to_s,
                                             :invoiceids => payload[:invoiceids].to_s,
                                             :notes => payload[:notes].to_s })

    logger.info "Project updated:  #{response}"
    return json(response)
  end

  # AddProjectMessage
  post "/project/message/add" do
    payload = JSON.parse(request.body.read.to_s)

    logger.info "Add message to Project: #{payload}"
    response = Whmcs.call("AddProjectMessage", { :projectid => payload[:projectid].to_i,
                                                 :message => payload[:message].to_s,
                                                 :adminid => payload[:adminid].to_i })

    logger.info "Message added:  #{response}"
    return json(response)
  end

  # AddProjectTask
  post "/task/add" do
    payload = JSON.parse(request.body.read.to_s)

    logger.info "Add Project Task: #{payload}"
    response = Whmcs.call("AddProjectTask", { :projectid => payload[:projectid].to_i,
                                              :duedate => payload[:duedate].to_s,
                                              :adminid => payload[:adminid].to_i,
                                              :task => payload[:task].to_s,
                                              :notes => payload[:notes].to_s,
                                              :completed => payload[:completed],
                                              :billed => payload[:billed] })

    logger.info "Task added:  #{response}"
    return json(response)
  end

  # UpdateProjectTask
  post "/task/update" do
    payload = JSON.parse(request.body.read.to_s)

    logger.info "Update Project Task: #{payload}"
    response = Whmcs.call("UpdateProjectTask", { :taskid => payload[:taskid].to_i,
                                                 :projectid => payload[:projectid].to_i,
                                                 :duedate => payload[:duedate].to_s,
                                                 :adminid => payload[:adminid].to_i,
                                                 :task => payload[:task].to_s,
                                                 :notes => payload[:notes].to_s,
                                                 :completed => payload[:completed] })

    logger.info "Task updated:  #{response}"
    return json(response)
  end

  # DeleteProjectTask
  get "/task/delete" do
    response = Whmcs.call("DeleteProjectTask", { :projectid => params[:projectid].to_i,
                                                 :taskid => params[:taskid].to_i })
    logger.info '[#{response[:result].upcase}] - #{response[:message]}.'
    return json(response)
  end

  # StartTaskTimer
  get "/timer/start" do
    response = Whmcs.call("StartTaskTimer", { :timerid => params[:timerid].to_i,
                                              :projectid => params[:projectid].to_i,
                                              :adminid => params[:adminid].to_i,
                                              #The start time as a unix time stamp.Defaults to time() if not provided
                                              :start_time => params[:start_time].to_i,
                                              :end_time => params[:end_time].to_i })
    logger.info '[#{response[:result].upcase}] - #{response[:message]}.'
    return json(response)
  end

  # EndTaskTimer
  get "/timer/end" do
    response = Whmcs.call("EndTaskTimer", { :timerid => params[:timerid].to_i,
                                            :projectid => params[:projectid].to_i,
                                            :adminid => params[:adminid].to_i,
                                            #The start time as a unix time stamp.Defaults to time() if not provided
                                            :end_time => params[:end_time].to_i })
    logger.info '[#{response[:result].upcase}] - #{response[:message]}.'
    return json(response)
  end
end
